pkg/controller/service/service_controller.go

```diff
@@ -636,9 +521,22 @@ func (s *ServiceController) nodeSyncLoop() {
 	glog.Infof("Detected change in list of current cluster nodes. New node set: %v",
 		nodeNames(newHosts))
 
-	// Try updating all services, and save the ones that fail to try again next
-	// round.
-	s.servicesToUpdate = s.cache.allServices()
+	// Try updating all services, and save the ones that fail to try again next round.
+	services, err := s.serviceLister.List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Failed to retrieve services from lister: %v", err)
+		return
+	}
+
+	// Filter out services that don't have load balancers already created since we can't update
+	// the hosts of a load balancer if the service doesn't have a load balancer.
+	s.servicesToUpdate = []*v1.Service{}
+	for _, svc := range services {
+		if len(svc.Status.LoadBalancer.Ingress) > 0 {
```

Comment:
Could we just check if type is `LoadBalancer` instead or do we feel that check is not strict enough?

Reply:
> updating the status is not something we force all providers to do.

What cloud doesn't provide an updated status after successfully calling `EnsureLoadBalancer`? I thought that was required.

Reply:
We could instead check if the finalizer exists & not being deleted instead since we can assume that those services do have a load balancer.

Reply:
It's not strictly required since we don't return/log errors (see [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L312-L327)). I think we should make it required but maybe outside the scope of this PR.

Reply:
After speaking offline, I remembered the reason for this was because the node sync loop calls [UpdateLoadBalancer](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L104) which not all clouds treat the same, so some require that the LB be created/up in order to call it. According to the API for service status, the status should exist if the load balancer exists, so this is the only way to ensure the load balancer is actually created.